Add blog post for `astro-theme-provider` (#19)
astrolicious/dotdev: apps/site/src/content/blog/astro-theme-provider.mdx

# Introducing: Astro Theme Provider

> ## What is Astro Theme Provider?

Reviewer: ```suggestion
```

I would get rid of this first heading.

Astro Theme Provider is a tool that allows you to author themes for Astro like a normal project and export your work as an integration for others to use. 

## Problems with current theme practices

In Astro, normally a theme is a full project published as a template that users can clone and customize. Project templates are a common pattern and have lots of use cases, but they also have downsides that theme authors may not always be aware of:

- **Hard to upgrade**: Merging improvements from a project template can be difficult if not impossible after customizing your local copy.
- **Incompatible with existing projects**: Project templates cannot be easily added to an existing Astro project. Adding a template to an existing project requires manuallly splicing code from the template into the project.
- **Not beginner-friendly**: The main demographic for theme users are people who are new to Astro and web development in general. Cloning and editing a project template as a beginner can be challenging. Project templates often times use lots of different technologies and can be bloated with files beginners may not need to see or touch.

> - **Inconsistent**: There is no standard for authoring templates, most templates use different combinations of patterns and structures. For example, many templates have a config file for customization, but the location of the file and its options vary depending on the template. Learning how to use a template while also learning how to use Astro can be confusing.

Reviewer: I would personally try to make this a paragraph, instead of bullet points, maybe try to link those sentences?

> ## A better way to create themes
> 
> The first step to making themes more user-friendly is shifting away from project templates and towards Astro integrations. Using an integration has many different advantages over a traditional template, for example:
> 
> * Publish and use themes as a NPM package
> * Update a theme by updating its package
> * Install themes using a single command \`astro add theme\`
> * Add themes to existing projects
> * Configure a theme by passing arguments
> 
> Although theme integrations are more user-friendly than project templates, it comes at the cost of being harder to create. They require a lot of boilerplate and a lower level knowledge of Astro and Vite making it inaccessible for a lot of developers.
> 
> Astro Theme Provider solves this by making it easier for theme authors to create theme integrations without having to write any integration code. Themes can be authored similar to a normal Astro project and be exported for others to use with a single function call.

Reviewer: ```suggestion
## A better way to create themes

The first step to making themes more user-friendly is shifting away from project templates and towards Astro integrations. Using an integration has many different advantages over a traditional template.

Theme integrations can be published as a package and used by users as a dependency, which also allows for easier updates. A big improvement is that themes can be installed with \`astro add theme\`. Additionally to this, users can add themes to existing projects, which decreases the time and effort it takes to add a theme or try different themes. Theme authors can make their theme configurable using simple settings. 

Although theme integrations are more user-friendly than project templates, it comes at the cost of being harder to create. They require a lot of boilerplate and a lower level knowledge of Astro and Vite making it inaccessible for a lot of developers.

Astro Theme Provider solves this by making it easier for theme authors to create theme integrations without having to write any integration code. Themes can be authored similar to a normal Astro project and be exported for others to use with a single function call.
```

I personally would try not to use bullet points here, this is my suggestions, you don't have to use it and can adapt it.

```
package/
├── public
├── src/
│   ├── assets
│   ├── components
│   ├── css
│   ├── layouts
│   └── pages
├── index.ts
└── package.json
```

```ts
// /package/index.ts
import { z } from 'astro/zod'; 
import defineTheme from 'astro-theme-provider';

export default defineTheme({
    schema: z.object({
        title: z.string()
    })
})
```

After creating your theme, it can be used inside an Astro project as an integration:

```ts
import { defineConfig } from "astro/config";
import myTheme from "my-theme";

export default defineConfig({
    integrations: [
        myTheme({
            config: {
                title: "Welcome to my theme!",
            },
        }),
    ],
});
```

> ## Using the right tool for the job
> 
> Theme integrations may be easier to use, but there are some cases where using a traditional project template makes more sense. For example, themes that require modifications to the source code, like a SAAS starter, are better suited as project templates. This can also make theme integrations harder to develop. Since users can't edit the source code themselves, authors have to create more verbose pages and components to support customizations from the user.
> 
> Ultimately, it is up to the theme author to weigh the advantages and disadvantages of both approaches and choose the one that works best for their project.

Reviewer: ```suggestion
```

I personally would get rid of this section completely, it is important to highlight that, but it should go in the README or the docs in the future,  I don't think it should go in the announcement blog post.

## Real world example

> The creation of Astro Theme Provider was inspired by Starlight, Astro's official documentation theme. Starlight is a theme integration that allows you to build a full documentation website with a config object and some markdown. Its ease of setup and use has made Starlight one of the most popular themes for Astro. Its success proves that theme integrations are a powerful pattern that have been under utilized in the Astro ecosystem.

Reviewer: ```suggestion
Astro Theme Provider is inspired by Starlight, Astro's official documentation theme. Starlight is a theme integration that allows you to build a full documentation website with a config object and some markdown. Its ease of setup and use has made Starlight one of the most popular themes for Astro. Its success proves that theme integrations are a powerful pattern that have been under utilized in the Astro ecosystem.
```

## What’s Next

> We are looking for help! Astro Theme Provider is in a beta stage, you can create themes with it, but it has not been officially released. The project is being opened up to the community so that others can get involved. If you are looking to help, you can check out the contributing guide. We encourage contributions of any kind! If you have any feedback or suggestions for the project, you can join the Discord channel or use GitHub Issues and Discussions.

Reviewer: ```suggestion
We're excited to share that Astro Theme Provider is in beta as of today. What's next? - We're working on improving the experience and shaping a stable version, but we need your help! We encourage contributions of any kind! You can check out the Project on [GitHub](https://github.com/astrolicious/astro-theme-provider) and join us on [Discord](https://chat.astrolicious.dev)!
```

I think you did a great job, but some sentences feel having the same point. Above my suggestion, but feel free to not use it, but make your own version :)
